src/cs/usfca/edu/histfavcheckout/controller/HistFavCheckoutHandler.java

```diff
@@ -168,17 +168,23 @@ else if(selected.equals("favs")) {
 		else if(selected.equals("ratings")){
 			users = userRepository.getTopRaters(PageRequest.of(page, nums));
 		}
-		LinkedHashMap<Integer, TopUser> topUserMap = new LinkedHashMap<Integer, TopUser>();
-		for(TopUser topUser : users) {
-			topUserMap.put(topUser.getUserId(), topUser);
+		else {
+			return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(
+					new OperationalResponse(false, "selected should be one of favs, checkouts or ratings"));
 		}
-		List<UserInfoResponse.UserInfo> userInfos = APIClient.getTopUsers(topUserMap.keySet());
+		LinkedHashMap<Integer, TopUser> topUserMap = new LinkedHashMap<Integer, TopUser>();
 		List<TopUserResponse> response = new LinkedList<TopUserResponse>();
-		for(UserInfoResponse.UserInfo userInfo : userInfos) {
-			response.add(new TopUserResponse(userInfo.getUserName(), 
-					userInfo.getEmail(), topUserMap.get(userInfo.getUserId()).getFavsCount(), 
-					topUserMap.get(userInfo.getUserId()).getCheckoutsCount(),
-					topUserMap.get(userInfo.getUserId()).getRatingsCount()));
+		if(!users.isEmpty()) {
+			for(TopUser topUser : users) {
+				topUserMap.put(topUser.getUserId(), topUser);
+			}
+			List<UserInfoResponse.UserInfo> userInfos = APIClient.getTopUsers(topUserMap.keySet());
+			for(UserInfoResponse.UserInfo userInfo : userInfos) {
+				response.add(new TopUserResponse(userInfo.getUserName(), 
+						userInfo.getEmail(), topUserMap.get(userInfo.getUserId()).getFavsCount(), 
+						topUserMap.get(userInfo.getUserId()).getCheckoutsCount(),
+						topUserMap.get(userInfo.getUserId()).getRatingsCount()));
+			}
```

Comment:
Nice work overall! Just wanted to point out something I noticed while testing.
I tried the url: https://hist-fav-checkout-beta.herokuapp.com/getTopUsers?selected=favs&page=0&nums=10

And I get back the response:
```
[
    {
        "userName": "nat",
        "email": "nat@gmail.com",
        "favsCount": 0,
        "checkoutsCount": 2,
        "ratingsCount": 1
    },
    {
        "userName": "testUser1",
        "email": "test1@hotmail.com",
        "favsCount": 1,
        "checkoutsCount": 1,
        "ratingsCount": 0
    },
    {
        "userName": "testUser4",
        "email": "test4@hotmail.com",
        "favsCount": 1,
        "checkoutsCount": 0,
        "ratingsCount": 0
    },
    {
        "userName": "testUser24",
        "email": "test24@hotmail.com",
        "favsCount": 0,
        "checkoutsCount": 0,
        "ratingsCount": 1
    },
    {
        "userName": "testUser46",
        "email": "test46@hotmail.com",
        "favsCount": 0,
        "checkoutsCount": 0,
        "ratingsCount": 1
    }
]
```

So the output isn't ordered by Favs count anymore because the first element has FavsCount of 0 and the next two elements have FavCount of 1.

I think this has something to do with line [182](https://github.com/usfcs-perf-eng-s20/heroku-project-favs/pull/28/files#diff-0fae860ae0c1e35be827d26bc884c318R182). The response coming back from User API is probably not in the same order as we expect, so it messes up the ordering.

I honestly haven't come up with a good approach to solve this too, but just wanted to mention the issue.

Reply:
Yes, I agree. It looks like Login team do not maintain the order. I have updated the CR accordingly and tested.